### ISSUE TITLE
Is it possible to combine channels and serialisation?

### ISSUE THREAD

**Issue by mikehearn** [puniverse/quasar#141]
It isn't clear to me how to use the serialisation framework. Imagine I have a fiber that's blocked waiting to receive something on a channel. Before blocking, I'd like to serialise the state of the fiber to checkpoint it, so if the app dies whilst the fiber is waiting, I can reload it from disk and restart it. But I can't combine the two as the serialisation support requires use of the lower level parkAndSerialise method.

**Comment by pron** [comment]
Fiber serialization is a very advanced, and quite experimental feature. It is experimental in the sense that we (and by that I mean the entire community of Quasar users) need to learn when it is most useful and how to use it. For now, it seems to me that fiber serialization should be best used with special constructs. One of them is actor migration. An actor can choose to suspend itself, and is then re-animated, possibly on another machine.

The scenario you've described sounds complicated it face, but may be useful. Suppose a fiber is serialized while being blocked on a channel. What happens to the channel when the fiber is serialized? What happens if the channel no longer exists when the fiber is deserialized? It seems to me that a channel is external state -- like a socket or any file descriptor -- that doesn't combine well with serialization. Do you have any ideas of how that would look like?

The best approach is probably for the fiber to suspend and serialize itself before receiveing from the channel (even if in a loop), and then, when deserialized, re-obtaining a reference to the channel through some mechanism.

**Comment by mikehearn** [comment]
For now I'd like to use the fibers framework without the rest of Quasar, with the option of using more of it later. Essentially I would like:

1. Take a fiber, run it with some sort of input object, do a blocking wait until it suspends itself.
2. Potentially serialise it now it's suspended.
3. Feed an object back in to resume it.

How the fiber interacts with the outside world and the like is something I need to handle myself right now rather than using the actors framework. Note that the serialisation here is something being handled outside of the fiber code itself.

Currently I have a bit of code that does this and works fine (I am happy with the API) using the old JavaFlow framework, but the continuations infrastructure in Quasar is much better. So I'm trying to figure out how to get a JavaFlow like API out of Quasar, to take advantage of the better maintained and more sophisticated bytecode instrumentation code here.

Does that make sense?

**Comment by pron** [comment]
So you'd just like to use fibers as continuations? You're not even interested in the scheduling (the scheduling is what makes the use of concurrency constructs possible)?

It is our plan to provide a `Continuation` class in Java 9, when Quasar no longer requires any manual annotation of suspendable methods (we're working with Oracle to make that happen), but so far we've found that unless you want to use continuations for some non-scheduling-related activity (i.e. nondeterminism, generators), fibers (i.e., *scheduled* continuations, are almost always preferable).

In any case, you can achieve what you want with a combination of a simple fiber scheduler that just uses the current thread as the executor (that should take you about one line of code using `FiberExecutorScheduler`), and then use the `Fiber.park` and `Fiber.unpark` primitives. For serialization/deserialization, use `Fiber.parkAndSerialize` and `Fiber.unparkSerialized`.

**Comment by mikehearn** [comment]
Concurrency might well come later but right now I want to use continuations to just simplify callback-ridden code, e.g. generator style.

Yes, I'm starting to see that. I think I may need to write a patch for the FiberAsync class, because parts of this code I am working with generate Guava ListenableFuture's and I'd often like to suspend the fiber until that future completes. I've got an adapter class that does that (no problem) but then, FiberAsync calls park and not parkAndSerialize. Does extending FiberAsync to support that sound reasonable? 

BTW what are you working on with Oracle? That sounds interesting.

**Comment by pron** [comment]
> because parts of this code I am working with generate Guava ListenableFuture's and I'd often like to suspend the fiber until that future completes

:) That's great, then, because Quasar comes prepackaged with `FiberAsync` support for `ListenableFuture`: [`AsyncListenableFuture`](http://docs.paralleluniverse.co/quasar/javadoc/co/paralleluniverse/fibers/futures/AsyncListenableFuture.html). All you have to do is call:

    AsyncListenableFuture.get(future)

and your fiber will block until the callback completes.

Also, `FiberAsync` is designed for extension; no need for patching. E.g. you can look at [`AsyncListenableFuture`'s code](https://github.com/puniverse/quasar/blob/master/quasar-core/src/main/java/co/paralleluniverse/fibers/futures/AsyncListenableFuture.java).

We're working on a new experimental JDK API for accessing the stack contents.

**Comment by mikehearn** [comment]
Ah! What a dummy I am. I wrote my own adapter and didn't realise one was already there. I'll get rid of mine and use the standard one.

Minor note: Guava has MoreExecutors.directExecutor() now which does the same thing as the SameThreadExecutor you define in that class.

I see what you mean - parkAndSerialize just registers a post-park action that then uses Kryo to serialise the fiber object. My code is also using Kryo extensively so this is convenient.  And FiberAsync just calls requestAsync inside its own ParkAction. 

Two things that aren't clear to me:

1. The fiber "current" passed to the ParkAction isn't given to the requestAsync method so I can't grab it and serialise it. It's not clear to me if there's another way to get "current" from inside requestAsync, or whether it does need a patch to just pass that as a parameter.

2. parkAndSerialize uses a thing called SERIALIZER_BLOCKER. I would not have access to that if I skipped parkAndSerialize and did it myself. Does that matter?

I may also be tempted to hack things a bit so I can provide my own Kryo, but it looks like your setup code is similar to my own, so perhaps it's not necessary.

**Comment by pron** [comment]
> The fiber "current" passed to the ParkAction isn't given to the requestAsync method so I can't grab it and serialise it

`FiberAsync` instances must be allocated by the same fiber that blocks on them, so you can grab `Fiber.currentFiber()` in the constructor and store it in a field for later use (that's how we do it).

> parkAndSerialize uses a thing called SERIALIZER_BLOCKER. I would not have access to that if I skipped parkAndSerialize and did it myself. Does that matter?

I don't think it does.

**Comment by mikehearn** [comment]
Thanks. I'll try that.

I don't see how better access to the stack contents would eliminate the need for manual annotations of the suspendable methods. I thought the point of the bytecode rewriting was partly to enable resumption at the right point, skipping all prior executed code. Short of full JVM support for continuations, that'd still be a requirement I'd have thought.

**Comment by pron** [comment]
Oh, there would still be a need for instrumentation, but that would allow us to figure out which methods to instrument *at runtime* while it's still not too late (i.e. after we no longer have access to the stack contents). A method will be instrumented the first time it blocks a fiber. We won't need to know about it in advance.

**Comment by mikehearn** [comment]
I see. I didn't realise it was possible to do bytecode rewriting on possibly live code like that.

**Comment by pron** [comment]
Oh, sure. The new version of a method (post-instrumentation) takes effect the next time it's entered, and you can repeat the process as many times as you like.

**Comment by mikehearn** [comment]
I hit another issue :) 

The framework tries to ensure that the blocker and unblocker objects are the same and silently refuses to unpark if they are not. Questions:

1. Shouldn't it throw an exception if it's going to refuse to unpark?
2. The parkAndSerialize method uses this mechanism to force you to unpark with unparkAndSerialize. But as I'm using serialization for checkpointing, I'd like to serialise but sometimes do a regular unpark, without using unparkDeserialized which seems to do some unnecessary stuff. For that I'd need access to the SERIALIZER_BLOCKER object but it's private.
3. What is the purpose of this mechanism?

I can work around this for now by just using reflection to grab the magic object. But I'd like to know why it works this way and the code comments weren't especially enlightening. Thanks!

**Comment by pron** [comment]
Oh, so you want to serialize and then continue? TBH, it's a use case I haven't considered.

In any event, blocker is usually for debugging purposes, but it is used when you want to ensure that only you would be able to unblock -- i.e. prevent spurious wakeups for whatever reason (e.g., if you can't easily access the condition variable, as in the case of `FiberAsync`). Throwing an exception is not what you want, as an `unpark` us usually a signal (a "lease") that says, "as far as I'm concerned, that fiber can be made runnable". 

You can either access the blocker with reflection, as you suggest, use `FiberTask.EMERGENCY_UNBLOCKER` which always works. 

If you have any suggestions on making the design of fiber serialization more amenable to various uses, I'd be happy to hear them. As I said before, it's pretty experimental, at least in the sense that we're not yet familiar with how people would want to use it. We're using it to transfer a fiber from one machine to another, so that's the only use-case we've considered.

**Comment by mikehearn** [comment]
I'm experimenting with using it for checkpointing. Consider a fiber/actor that takes part in a complex multi-step protocol, potentially with several remote nodes. I would like to be able to kill the hosting process at any point, restart it, and have those actors resume. With a reliable message layer and serialisation of fibers to checkpoints whenever they send/block waiting for a message (or alter the outside world in some other way), it seems like it could be close to transparent, with some caveats around versioning across upgrades. That's my current (exploratory) use case.

**Comment by mikehearn** [comment]
Yeah, so, I think I'm going to have to branch quasar and start submitting pull requests for these things, the number of workarounds keeps going up.

Attempting to unpark a fiber that was parked with parkAndSerialize when using a SameThreadExecutor crashes the JVM. The issue is the unsafe tricks done to swap ThreadLocals around. It ends up setting the TLS table to null when there are active ThreadLocals (like currentStrand) that expect it to be there. Not sure why.

This doesn't happen if unparkAndSerialize is used.

**Comment by mikehearn** [comment]
Actually it seems, even when matching parkAndSerialise with unparkSerialised, it can still fail, I just wasn't hitting a codepath that was accessing TLS slots in my separate test program. Enabling debugMode makes it crash again. The swapping logic clearly doesn't expect everything to be running on a single thread.

**Comment by pron** [comment]
Would you like to open an issue for that?

**Comment by mikehearn** [comment]
Thanks for merging the fixes! I don't suppose it'd be possible to refresh the 0.7.4 snapshot would it? Then I can push my code that relies on the fixes to other developers without asking them to install a custom build of Quasar.

**Comment by pron** [comment]
Done.
BTW, we should be cutting a release within a week.

**Comment by mikehearn** [comment]
This seems to be working OK now, at least in a unit testing environment, so I'll close this issue. Thanks for all your help.

**Comment by pron** [comment]
Brilliant!
